mgr/dashboard: asynchronous task support

rjfd opened the pull request:
This PR introduces the support for executing long-running tasks by dashboard backend controllers.
The tasks are executed asynchronously and can be queried for their respecting executing status.

For detailed information please read the changes made to `HACKING.rst` file.

Signed-off-by: Ricardo Dias <rdias@suse.com>

votdev commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,159 @@ The settings management implementation will make sure that if you change a
+
+To help in the development of the above scenario we added the support for
+asynchronous tasks. To trigger the execution of an asynchronous task we must
+use the follwoing class method of the ``TaskManager`` class::
```
s/follwoing/following/

votdev commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,159 @@ The settings management implementation will make sure that if you change a
+* ``func`` is the python function that implements the operation code, which
+  will be executed asynchronously.
+
+* ``args`` and ``kwargs`` are the positional and named argguments that will be
```
s/argguments/arguments/

votdev commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,159 @@ The settings management implementation will make sure that if you change a
+is not created and you get the task object of the current running task.
+
+
+How to get the list of executing and finished asynchronous tasks?
```
Maybe 'running' is better than 'executing' here.

ricardoasmarques commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+
+To help in the development of the above scenario we added the support for
+asynchronous tasks. To trigger the execution of an asynchronous task we must
+use the follwoing class method of the ``TaskManager`` class::
```
s/follwoing/following/

ricardoasmarques commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+asynchronous tasks. To trigger the execution of an asynchronous task we must
+use the follwoing class method of the ``TaskManager`` class::
+
+  import ..tools import TaskManager
```
s/import ..tools import TaskManager/from .tools import TaskManager/

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+``value == None``, and if ``state == VALUE_EXCEPTION`` then ``value`` stores
+the exception object raised by the execution of function ``func``.
+
+The pair ``(namespace, metadata)`` should univocally identify the task being
```
why not use a task name instead? Why invent something like `namespace`?

rjfd replied:
`(namespace, metadata)` provides the ability to identify unequivocally a task without having to know the "task name" or "task id". This provides much more flexibility for the consumers of the list of executing/finished tasks, as they are not require to store the list of task names that they want keep track.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+
+* ``VALUE_DONE = 0``
+* ``VALUE_EXECUTING = 1``
+* ``VALUE_EXCEPTION = 2``
```
Please remove `VALUE_EXCEPTION` as a valid return code. Instead raise the original exception in the caller's thread. If someone wants to catch a specific exception, he should use `try: ... except:' instead.

rjfd replied:
Right, maybe we don't need to pass the exception this way. Will try your suggestion.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+The ``value`` will store the result of the execution of function ``func`` if
+``state == VALUE_DONE``. If ``state == VALUE_EXECUTING`` then
+``value == None``, and if ``state == VALUE_EXCEPTION`` then ``value`` stores
+the exception object raised by the execution of function ``func``.
```
Remove this last sentence.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+The pair ``(namespace, metadata)`` should univocally identify the task being
+run, which means that if you try to trigger a new task that matches the same
+``(namespace, metadata)`` pair of the currently running task, then the new task
+is not created and you get the task object of the current running task.
```
This is bad. Consider:
```
def set_pg_count_per_pool(pool, count):
    CephService.send_message(....)
   ...

TaskManager.run('set_pg_count_per_pool', 42, lambda: set_pg_count_per_pool(mypool, 42))
TaskManager.run('set_pg_count_per_pool', 12, lambda: set_pg_count_per_pool(mypool, 12))
TaskManager.run('set_pg_count_per_pool', 42, lambda: set_pg_count_per_pool(mypool, 42))
```

In the end, the pool would have 12 pgs, instead of 42. But the user would have a successful answer from the UI. 

Instead, `TaskManager` should *per default* create every task independent of any parameters.

rjfd replied:
First, the developer needs to follow the contract of `TaskManager.run` as stated in `HACKING.rst`, which says that `(namespace, metadata)` must unequivocally identify a task, and as also stated in the documentation, he knows that if he triggers a new task with the same `(namespace, metadata)` as an already executing task that he will not be triggering a new task.
Therefore, according to the documentation the developer should not write the code you used as an example.

Moreover, the above example is a good example of a bad specification of the `(namespace, metadata)`. For the above example, the `(namespace, metadata)` could be: `("osd/pool/set_pg_count", {"pool_name": mypool})`. This way, the frontend could easily map the executing tasks to the "pool" table in the "Pools Page" (that does not exist yet).

sebastian-philipp replied:
> Moreover, the above example is a good example of a bad specification of the `(namespace, metadata)`. For the above example, the `(namespace, metadata)` could be: `("osd/pool/set_pg_count", {"pool_name": mypool})`. This way, the frontend could easily map the executing tasks to the "pool" table in the "Pools Page" (that does not exist yet).

Your example supports my arguments. `("osd/pool/set_pg_count", {"pool_name": mypool})` will make things even worse by silently ignoring _all_ `set_pg_count` tasks while the first one is still running. Which means that a developer has to be careful in designing a task's metadata, in order to not run into the trap of being silently ignored. This shouldn't be necessary and looks to me like a bug.

Finally, `"osd/pool/set_pg_count"` is exactly something that I would think of a perfect example of a task name: it consists of the controller class name and the method name.

rjfd replied:
@sebastian-philipp you can't use nonsensical example as an argument. Why would the user issue three asynchronous "set_pg_count" operations targeting the same pool?

The `("osd/pool/set_pg_count", {"pool_name": mypool})` specification actually avoids someone from doing such mistake, by guaranteeing that only one "set_pg_count" operation is running at a time.

jecluis replied:
I'm not entirely following what the divisiveness here is. Correct me if I'm wrong:

1. We have a queue of tasks being executed or pending execution, identified by a given key consisting of `(namespace, metadata)`.

2. A developer queuing one task `(foo/bar, 1337)` will be able to later check on said task using the previously defined key.

3. A developer queuing multiple tasks `(foo/bar, 1337)` and `(foo/baz, 1337)` will in fact be queuing two different tasks, which can be checked up using their individual keys.

4. A developer queuing multiple tasks `(foo/bar, 1337)` will in fact be queuing ONLY one task.

5. Tasks being queued with the same key `(foo/bar, 1337)` while an already existing task with said key is executing, will be considered idempotent requests and will end up waiting for the same, original task to finish.

6. A developer checking up on a task `(foo/bar, 1337)` multiple times will always get the same response from the queue, whether it's still executing or finished execution, until a given number of seconds/minutes has passed.

Am I missing something?

Even although I haven't read the code in its entirety, the architectural aspect of how this works makes a lot of sense to me. Surely a developer will be able to abuse it, but that's why critical code review should be enforced.

LenzGr replied:
I'd like to get @ricardoasmarques take on this, as he's currently working on the implementation that will have to consume this API. We should make sure that his requirements are being met and addressed.

ricardoasmarques replied:
From the consumer/front-end point of view, it makes sense to have a unique identifier composed by `namespace` and `metadata` (we can call it `name` or something else if you prefer to have a different name, but the behaviour should remain the same).

If I understand correctly, to follow @sebastian-philipp suggestion, we will need to have a unique identifier generated by server (e.g. a unique numeric id), and we don't want to depend on that numeric id on the frontend because this will make it very hard to deal with cases where user do a full page refresh or move to a diferent browser during the execution of a task.


Taking into account the experience we had with openATTIC while implementing iSCSI and NFS service management, we simply ignored the numeric taskid on the frontend when displaying the "Starting..." and "Stopping..." messages to guarantee that the current status is always displayed even after a full page refresh.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+and finished tasks.
+
+The ``Task`` controller exposes the ``/api/task`` endpoint that returns the
+list of executing and finished tasks. This endpoint accepts the ``namespace``
```
> list of executing and finished tasks

There are dozens of places where you differentiate between executing and finished tasks. Instead, I'd keep all tasks  in one list and add a `state` property to each task.

rjfd replied:
I think having them separate is easier for the consumers. I believe the executing tasks list will be the most used, but will have less entries than the finished tasks list. Also, I don't think consumers will process both kinds of tasks at the same time, therefore it's more beneficial to have them separate.

But maybe some of the frontend guys can give their opinion on this as they will be the main consumers: @ricardoasmarques @tspmelo @votdev ?

sebastian-philipp replied:
I'm especially talking about the backend implementation. The API should be of course designed to be easily consumable by the front end. But keep in mind that the current task queue dialog does accept them as a single list of entries

rjfd replied:
@sebastian-philipp what current task queue dialog are you talking about?

sebastian-philipp replied:
@rjfd http://demo.openattic.org/openattic/ one of the icons in the top right corner

ricardoasmarques replied:
ATM I see 2 scenarios where this list(s) will be used by the frontend.

1) To display the status of an item, example: "asynchronous RBD deletion"

![screenshot from 2018-03-14 13-11-10](https://user-images.githubusercontent.com/14297426/37404270-4b78e9f8-2789-11e8-874b-aac9d1b65ddb.png)

To implement this scenario , we just need to use the "executing_tasks" list, "finished_tasks" can be ignored here.

2) To display background tasks

![screenshot from 2018-03-14 13-06-42](https://user-images.githubusercontent.com/14297426/37404375-94d9d8be-2789-11e8-8544-1040ac5e35a0.png)

If this prototype (that I'm currently working on) is accepted, having 2 lists is useful.

-----

So, in both scenarios having 2 lists matches the front-end requirements. Of course, if the back-end returns a single list, we can split it in the front-end.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+``rbd/``.
+
+To prevent the finished tasks list from growing unboundly, the finished tasks
+will be maintained in memory for 1 minute. After a minute, when the finished
```
Too short. The absolute minimum for a finished task to stay in the list would be a coffe-break. I'd keep them at least a day.

rjfd replied:
When I thought about how the "finished tasks" list would be used I reached a conclusion that they might not even be necessary, but I decided to keep them anyway and for that reason I choose a very small TTL. For instance, if you trigger a task to delete an RBD image, when the task finishes the image will immediately disappear from the RBD images table, therefore the user will know that the task is finished, without consulting the list of finished tasks.

But I don't mind to change the TTL value to something else as long as we have a quorum of people deciding on the same value.

sebastian-philipp replied:
Looks like we have different tasks in mind. "osd/pool/set_pg_count" for example may take quite some time. And as soon as it finishes, it would vanish too soon.

rjfd replied:
As I said before I don't think finished tasks will add much more information to the user, also they are only stored in ceph-mgr memory, which means that the frontend must cope with the fact that a finished task information may not be available.

But I agree that 1 minute might be too small. @LenzGr @ricardoasmarques what value for finished tasks TTL do you think it's appropriate, 1 day?

ricardoasmarques replied:
IMO, we should always maintaining in memory the last 20(?) finished tasks, and only use the TTL (1m?) if we have more than 20(?) newer tasks.

This can be useful to list the last finished tasks on the front-end because the end user does not expect to have finished tasks on the list that suddenly disappear, but he can easily understand that the oldest finished tasks are no longer visible.

rjfd replied:
@ricardoasmarques that's a good idea, and easy to implement. @sebastian-philipp what do you think?

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+
+  {
+    'namespace': "namespace",  # str
+    'metadata': { },  # dict
```
do you force a `dict` here? I'd just use the parameter provided to `TaskManager.run`

rjfd replied:
The `TaskManager.run` metadata parameter should be a dictionary, that way you can specify precise/useful information about the task. Of course the code will also work with other sort of data, but that is not advised.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+  {
+    'namespace': "namespace",  # str
+    'metadata': { },  # dict
+    'begin_time': 0.0,  # float
```
Always return dates as [ISO 8601](https://stackoverflow.com/a/15952652/3185053) in JSON.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+    'namespace': "namespace",  # str
+    'metadata': { },  # dict
+    'begin_time': 0.0,  # float
+    'progress': 0  # int (percentage)
```
I'd allow a float here.

rjfd replied:
Why do you need a float to represent the progress?

sebastian-philipp replied:
There is no requirement to use float. I just don't see a reason to force an integer heere

jecluis replied:
As, I'm guessing, there's no requirement to force a float? ;)

I'm good either way, but this interaction was rather amusing.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+  {
+    'namespace': "namespace",  # str
+    'metadata': { },  # dict
+    'begin_time': 0.0,  # float
```
again, ISO 8601

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+    'namespace': "namespace",  # str
+    'metadata': { },  # dict
+    'begin_time': 0.0,  # float
+    'end_time': 0.0,  # float
```
ISO 8601

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+    'end_time': 0.0,  # float
+    'latency': 0.0,  # float
+    'progress': 0  # int (percentage)
+    'success': True,  # bool
```
I'd replace with `state` add unify executing and finished tasks. This would also improve the TaskManager itself.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+    'latency': 0.0,  # float
+    'progress': 0  # int (percentage)
+    'success': True,  # bool
+    'ret_value': None,  # object, populated only if 'success' == True
```
I'd allow generic JSON data structures here. Sometimes, a simple Boolean is enough.

rjfd replied:
When I say `object`, I mean anything, it could be a boolean. Of course the restriction is that it must be something that can be serialized in JSON.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/controllers/task.py:
```diff
@@ -0,0 +1,15 @@
+        return {
+            'executing_tasks': executing_t,
+            'finished_tasks': finished_t
+        }
```
I'd unify both lists into one and add a `state` property.

jecluis replied:
I can imagine this being useful from the UI point-of-view. Maybe a frontend dev could opinionate on what they would like to be returned here?

rjfd commented on src/pybind/mgr/dashboard_v2/controllers/task.py:
```diff
@@ -0,0 +1,15 @@
+@ApiController('task')
+@AuthRequired()
+class TaskController(RESTController):
+    def list(self, namespace=None):
```
Namespaces are useful to group related tasks. For instance, to group all tasks related to a single component like RBD `rbd/*`.
Since the frontend is the main entity to consume the list of tasks, working with a "task name" is difficult. The frontend memory is volatile and might easily lose the task name upon refresh, therefore it is easier to use namespaces.

Also, you can define a unique namespace to mimic the same behavior of a "task name" or "task id".

The `namespace` parameter in this function accepts a glob expression, which makes it pretty good for filtering groups of tasks.

sebastian-philipp replied:
Just renaming `namespace` to `name` or `task_name` would make things clearer. Especially when moving the name to the task definition. We're talking about the same thing here.

jecluis replied:
I don't have strong feelings either way, but `namespace` does sound a bit better than `name` or `task_name` IF `namespace` here is supposed to allow more than just a specific task unique identifier; `name` and `task_name`, on the other hand, seem to convey something a bit too specific.

However, I don't particularly like `namespace` for this particular case, but I don't have a better name for it, so...

rjfd commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+      @cherrypy.expose
+      @cherrypy.tools.json_out()
+      def default(self):
+          task = TaskManager.run("dummy/task", {}, self._dummy)
```
Can you be more clear on what you mean by `should be part of the definition of the task`?

The `namespace` parameter is supposed to be a dynamic property on purpose, if the controller developer wants all its tasks to have the same `namespace` string then it can declare a constant variable with that string.

sebastian-philipp replied:
The task name is typically very similar to the method name plus class or package name. There is no need to make this a dynamic property, except if you have a clear usecase in mind.

jecluis replied:
@sebastian-philipp If you only have one single `task_name` as the sole identifier of a single task in the queue, how do you expect to handle different tasks performing the same base action if you are not also differentiating them based on their arguments?

You actually gave a pretty good example of how what you are proposing would not work (in another comment), even though you presented it as an argument against the current implementation:

> TaskManager.run('set_pg_count_per_pool', 42, lambda: set_pg_count_per_pool(mypool, 42))
> TaskManager.run('set_pg_count_per_pool', 12, lambda: set_pg_count_per_pool(mypool, 12))
> TaskManager.run('set_pg_count_per_pool', 42, lambda: set_pg_count_per_pool(mypool, 42))

Whereas what is above would be expected to run two different tasks, and have the third call being returned the status of the first call (given it's an idempotent operation), how would you handle such a sequence of operations if you only had one single, static identifier `task_name`?

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/tools.py:
```diff
@@ -521,11 +534,207 @@ def run(self):
+        with self.lock:
+            if self.executor_thread is None:
+                self.executor_thread = AsyncTask.ExecutorThread(self)
+                self.executor_thread.start()
```
You're unconditionally creating a new posix thread per scheduled task here and dashboard_v2 already starts about ten posix threads.

LenzGr replied:
But wouldn't each task have to be handled in a separate thread? How else would you execute them in parallel?

rjfd replied:
@LenzGr I think what @sebastian-philipp means is that with the current implementation the number of threads might grow unbounded. I agree that we can improve this part of code by using a thread queue that limits the number of worker threads. I think this improvement can be done in a separate PR and is not critical at this point.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/tools.py:
```diff
@@ -521,11 +534,207 @@ def run(self):
+                val = self._task.fn(*self._task.fn_args, **self._task.fn_kwargs)
+                self._task.end_time = time.time()
+            except Exception as ex:
+                logger.exception("Error while calling %s: ex=%s", self._task, str(ex))
```
replace with
```python
logger.exception("Error while calling %s", self._task)
```
as the exception is already printed by `logger.excepton`

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/tools.py:
```diff
@@ -521,11 +534,207 @@ def run(self):
+                    self._task.exception = ex
+            else:
+                with self._task.lock:
+                    self._task.latency = self._task.end_time - self._task.begin_time
```
`s/latency/duration`?

jecluis commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+
+* ``namespace`` is a string that can be used to group tasks. For instance
+  for RBD image creation tasks we could specify ``"rbd/create"`` as the
+  namespace, or conversly ``"rbd/remove"`` for RBD image removal tasks.
```
s/conversly/conversely/. Also, I don't think `conversely` applies here. I'd go with `similarly` instead, as you are not pointing out a reversion of the previous statement.

jecluis commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+The pair ``(namespace, metadata)`` should univocally identify the task being
+run, which means that if you try to trigger a new task that matches the same
+``(namespace, metadata)`` pair of the currently running task, then the new task
+is not created and you get the task object of the current running task.
```
I would also mention the idempotent nature of running two calls to `run()` with the same namespace and metadata, with an example.

jecluis commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+will return all executing and finished tasks which namespace starts with
+``rbd/``.
+
+To prevent the finished tasks list from growing unboundly, the finished tasks
```
s/unboundly/unbounded/

jecluis commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+  }
+
+
+How to updated the execution progress of an asynchronous task?
```
Either s/updated/update/, or you the verb.

jecluis commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+The ``inc_progress`` method receives as argument an integer value representing
+the delta we want to increment to the current execution progress percentage.
+
+Now we show a full example of a controller that triggers a new task and
```
Instead of

`Now we show a full example of a controller [...]`

I'd go with

`Take the following example of a controler [...]`

rjfd commented:
@votdev @ricardoasmarques @jecluis I addressed your documentation comments.

@sebastian-philipp I addressed all the comments that are not under discussion.

rjfd commented:
jenkins retest this please

rjfd commented:
@ricardoasmarques I implemented the solution you proposed for the clean up of old finished tasks.

rjfd commented:
jenkins retest this please

jcsp commented on src/pybind/mgr/dashboard/tools.py:
```diff
@@ -521,11 +534,217 @@ def run(self):
+    def _run(self):
+        with self.lock:
+            if self.executor_thread is None:
+                self.executor_thread = AsyncTask.ExecutorThread(self)
```
Threads are needed for tasks that use blocking apis, but I'm not sure about using them for tasks in general.

For tasks that use mon commands, it's nice to avoid the thread and just call back into them when the command response messages come in -- that way it's a lot easier to implement cancelling or otherwise interrupting the task.  That's handy for putting a cancel button in the UI, but I think we're also going to want it for cases where we might want to cancel some tasks in response to health issues.

rjfd replied:
When I first started thinking on the solution for the asynchronous tasks, I thought about this problem, but then when I was developing the solution I completely forgot to add the support for calling functions that were already asynchronous. Thanks for raising the issue.

My idea is to keep the `TaskManager` interface, so from the API consumers perspective they will only see "tasks". I think I already know how to implement the support for async API calls without changing too much the current interface. I'll push the implementation to this PR and then notify you for the review.

rjfd commented on src/pybind/mgr/dashboard/HACKING.rst:
```diff
@@ -508,3 +508,171 @@ The settings management implementation will make sure that if you change a
+  # ...
+  TaskManager.run(namespace, metadata, func, args, kwargs)
+
+* ``namespace`` is a string that can be used to group tasks. For instance
```
The reason I called it namespace was to give a hint to the developer that it should use a string with some structured meaning. This string will be used by API consumers (currently only the frontend) and it shouldn't reflect any implementation details of the backend. For instance, even if we refactor the backend code, the frontend code should keep working. That's why I'm a bit reluctant to call it something with "function" prefix in the parameter name. Maybe I'll just call it `name`, does that address your concern?

jcsp replied:
`name` sounds good to me (sorry for delay responding)

rjfd replied:
renaming done

rjfd commented on src/pybind/mgr/dashboard/HACKING.rst:
```diff
@@ -508,3 +508,171 @@ The settings management implementation will make sure that if you change a
+The return value of the ``AsyncTask.wait`` method is a pair ``(state, value)``
+where ``state`` is an integer with following possible values:
+
+* ``VALUE_DONE = 0``
```
agreed, I'll change that to string instances.

rjfd commented on src/pybind/mgr/dashboard/tools.py:
```diff
@@ -521,11 +534,217 @@ def run(self):
+        return executing_tasks, finished_tasks
+
+    @classmethod
+    def list_serialized(cls, ns_glob=None):
```
I agree

rjfd commented:
@jcsp I've implemented the support for calling functions that are already non-blocking, you can check the details of how to call those functions within the asynchronous task infrastructure in the changes to the `HACKING.rst`.

@ricardoasmarques @jecluis @sebastian-philipp @votdev sorry, but I need a second round of reviews since I've rewritten a big part of the implementation. The interface for the developer remains the same as it was before apart the addition of custom task executors (see `HACKING.rst` for details)-

sebastian-philipp commented:
@rjfd How would you implement a task that monitors the successful creation of PGs? ([This is the asynchronous implementation of that task in openATTIC](https://bitbucket.org/openattic/openattic/src/ce4543d4cbedadc21b484a098102a16efec234f9/backend/ceph/tasks.py?at=master&fileviewer=file-view-default#tasks.py-47:77)). Would you start an extra thread for this task?

rjfd commented:
@sebastian-philipp here's an example how to mimic the oA code you referred to using dashboard asynchronous tasks. Since getting the `pg_status` of a pool is a blocking operation we will be executing the `monitor_pg_state` function as a task using the default `ThreadedExecutor`, which internally spawns a thread.

```python

def monitor_pg_state(pool_name, pg_num):
    pool_info = CephService.get_pool_info(pool_name)
    if not pool_info or 'pg_status' not in pool_info:
        return {'success': False, 'msg': "Could not get pool pg_status"}
    pg_status = pool_info['pg_status']
    active = 0
    if 'active+clean' in pg_status:
        if pg_status['active+clean'] == pg_num:
            return {'success': True, 'msg': "All PGs are active+clean"}
        else:
            active = pg_status['active+clean']

    progress = int(round(active * 100.0 / pg_num))
    TaskManager.current_task().set_progress(progress)

    time.sleep(1.0)  # sleep for a bit, no need to check every millisecond
    return monitor_pg_state(pool_name, pg_num)


@ApiController('test')
class Test(BaseController):
    @cherrypy.expose
    @cherrypy.tools.json_out()
    def pg_monitor(self, pool_name):
        pool_info = CephService.get_pool_info(pool_name)
        task = TaskManager.run("osd/pool/pg/monitor", {'pool_name': pool_name},
                               monitor_pg_state, [pool_name, pool_info['pg_num']])
        return task.wait(2.0)
```

rjfd commented:
@sebastian-philipp (and others @jecluis @ricardoasmarques @jcsp ) here's an example of how to implement the "create pool" operation using the asynchronous nature of `mgr.send_command` and a custom task executor. The implementation is based on an asynchronous state machine that executes all steps asynchronously, and the task only finishes when all pools PGs are in `active+clean` state.

https://github.com/rjfd/ceph/blob/wip-dashboard-pr-20865/src/pybind/mgr/dashboard/controllers/pool.py

rjfd commented:
Added cc5ef6a to this PR to avoid memory leaks because of dangling notificationqueue listeners registered by short lived objects like in the custom executor example in `HACKING.rst`. Updated the example accordingly.

ricardoasmarques commented:
@rjfd I've retested this PR and it still lgtm.

sebastian-philipp commented:
I just looked at `/task` and I'm still find it confusing to have two `success` bools contradicting each other:
```json
{
  "executing_tasks": [],
  "finished_tasks": [
    {
      "exception": null,
      "end_time": "2018-03-23T12:30:37.318413Z",
      "success": true,
      "begin_time": "2018-03-23T12:30:37.317111Z",
      "duration": 0.0013020038604736328,
      "progress": 100,
      "ret_value": {
        "msg": "Could not get pool pg_status",
        "success": false
      },
      "namespace": "osd/pool/pg/monitor",
      "metadata": {
        "pool_name": ".rgw.root"
      }
    }
  ]
}
```
But I'm OK with it, if @ricardoasmarques is handling this in the UI.

sebastian-philipp commented:
The first on is on the task and the second one is on the return value of the task.

rjfd commented:
Another round of QA tests were successful: http://pulpito.ceph.com/rdias-2018-03-27_15:04:07-rados:mgr-wip-rdias-testing-distro-basic-smithi/
